Improvement/web api framework optimization

NielsdeBoerDAE opened the pull request:
This pull request features the following improvements:

- ApiLogs table now properly logs the time taken
- Sql filters are now supported for SQL backends
- Implemented pagination (using ?offset=20&limit=20 for example). Keep in mind that because data dictionaries do not support offsets it still has to actually find the record. It just does not process it.
- General speed optimalizations. Certain attributes are now only fetched once from the database instead of for every single record
- Reuse json objects were possible for speed improvements
- reverse proxy's now supported
- Find index for cRestChildCollection is now a property
- Errors are now logged to the event log
- Fields that have a precision are now marked as number fields in the OpenApi specification instead of integer fields
- GET requests now show that they return an array instead of an object. Better reflecting the actual response.

Closes #15

NielsdeBoerDAE committed:
Improved label finding

Previously certain attributes were fetched from the database for every single record. Now these properties are defined on object initialization and only the property is retrieved for each find. Should give a pretty good performance increase.

NielsdeBoerDAE committed:
Improved label finding

Previously certain attributes were fetched from the database for every single record. Now these properties are defined on object initialization and only the property is retrieved for each find. Should give a pretty good performance increase.

NielsdeBoerDAE committed:
Implemented sql constrains

Applications that support sql filters can now use them to optimize constraints a little.

NielsdeBoerDAE committed:
Implemented sql constrains

Applications that support sql filters can now use them to optimize constraints a little.

NielsdeBoerDAE committed:
Turn find index for cRestChildCollection into a property

Allow developers to determine what index they want to use for finds on a child table.

NielsdeBoerDAE committed:
Turn find index for cRestChildCollection into a property

Allow developers to determine what index they want to use for finds on a child table.

NielsdeBoerDAE committed:
Allow query params to be passed twice

It is now possible to pass query parameters twice. This allows you to apply two types of constraints to the same column

NielsdeBoerDAE committed:
Allow query params to be passed twice

It is now possible to pass query parameters twice. This allows you to apply two types of constraints to the same column

NielsdeBoerDAE committed:
Optimized json handles

No longer create and destroy handles during a for loop. The handle will be created once and then be reused with InitializeJsonObject again. This should help improve performance

NielsdeBoerDAE committed:
Optimized json handles

No longer create and destroy handles during a for loop. The handle will be created once and then be reused with InitializeJsonObject again. This should help improve performance

NielsdeBoerDAE committed:
Implement number field for openapi specification

NielsdeBoerDAE committed:
Implement number field for openapi specification

NielsdeBoerDAE committed:
Log errors to event log

NielsdeBoerDAE committed:
Reverse proxy support for servers

NielsdeBoerDAE committed:
Reverse proxy support for servers

NielsdeBoerDAE committed:
Implement pagination

Implemented pagination for the cRestDataset. It is now possible to pass a offset in the query parameters. Since data dictionaries don't actually support an offset it still has to find the rows, it just skips over the first few depending on the offset.

NielsdeBoerDAE committed:
Implement pagination

Implemented pagination for the cRestDataset. It is now possible to pass a offset in the query parameters. Since data dictionaries don't actually support an offset it still has to find the rows, it just skips over the first few depending on the offset.

NielsdeBoerDAE committed:
GET requests should show that they return an array and not a single object

NielsdeBoerDAE committed:
GET requests should show that they return an array and not a single object

DataFlex-dev commented on Web API Library/AppSrc/WebApi/cRestDataset.pkg:
```diff
@@ -18,11 +18,17 @@ Class cRestDataset is a cBaseRestDataset
+        // This determines how many records we return during a get all
         Property Integer piLimitResults 0
+        // This determines how many records to skip during a GET request.
+        Property Integer piOffset 0
```
This should be runtime specific and not in a design-time object-class.

DataFlex-dev commented on Web API Library/AppSrc/WebApi/cRestDataset.pkg:
```diff
@@ -18,11 +18,17 @@ Class cRestDataset is a cBaseRestDataset
+        
+        // Grabs the old psSQLFilter and restores it
+        { Visibility=Private }
+        Property String psOldSQLfilter ""  
```
This should be runtime specific and not in a design-time object-class.

DataFlex-dev commented on Web API Library/AppSrc/WebApi/cOpenApiSpecification.pkg:
```diff
@@ -77,15 +78,25 @@ Class cOpenApiSpecification is a cObject
         If (sApiRoot = "") Begin
             Get ServerVariable of ghoWebServiceDispatcher "SERVER_NAME" to sServerName
             Get ServerVariable of ghoWebServiceDispatcher "URL" to sApiRoot
```
In the future see if we can check a forwarded for server name and url as well.

NielsdeBoerDAE committed:
Improve error handling while debugger is present

NielsdeBoerDAE committed:
Revert "Improve error handling while debugger is present"

This reverts commit ff1fe5960a25aa255787a9f630ab0e1956d1b5ff.

NielsdeBoerDAE committed:
Revert "Improve error handling while debugger is present"

This reverts commit ff1fe5960a25aa255787a9f630ab0e1956d1b5ff.

NielsdeBoerDAE committed:
Updated gitignore

NielsdeBoerDAE committed:
Improve sample

BasicAuth no longer creates sessions, this way requests are actually stateless. Additionally renamed the routers to have more descriptive names.

NielsdeBoerDAE committed:
piOffset and psSqlOldFilter are no longer design time properties

offset and oldsqlfilter are now part of the webapicallcontext and only part of the current request.

NielsdeBoerDAE committed:
Add templates to the library

Add the following templates to the library:

- cWebApi
- cWebApiRouter
- cRestDataset
- cRestField
- cRestEntity
- cRestChildCollection
- cWebApiModifier
- cWebApiAuthModifier
